mongo/datadog_checks/mongo/mongo.py

```diff
@@ -544,113 +244,54 @@ def total_seconds(td):
-
-        if self.do_auth:
-            if self.auth_source:
-                msg = "authSource was specified in the the server URL: using '%s' as the authentication database"
-                self.log.info(msg, self.auth_source)
+            if self.do_auth:
+                self.log.info("Using '%s' as the authentication database", self.auth_source)
                 self._authenticate(cli[self.auth_source])
-            else:
-                self._authenticate(db)
-
-        try:
-            status = db.command('serverStatus', tcmalloc=self.collect_tcmalloc_metrics)
         except Exception:
-            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL, tags=self.service_check_tags)
+            self.service_check(SERVICE_CHECK_NAME, AgentCheck.CRITICAL, tags=self.service_check_tags)
             raise
-        else:
-            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.OK, tags=self.service_check_tags)
 
-        if status['ok'] == 0:
-            raise Exception(status['errmsg'].__str__())
+        tags = deepcopy(self.base_tags)
 
-        ops = db.current_op()
-        status['fsyncLocked'] = 1 if ops.get('fsyncLock') else 0
+        self.update_deployment_type(cli['admin'])
+        if isinstance(self.deployment_type, ReplicaSetDeploymentType):
+            tags.extend(
+                [
+                    "replset_name:{}".format(self.deployment_type.replset_name),
+                    "replset_state:{}".format(self.deployment_type.replset_state_name),
+                ]
+            )
 
-        status['stats'] = db.command('dbstats')
-        dbstats = {self.db_name: {'stats': status['stats']}}
         try:
             mongo_version = cli.server_info().get('version', '0.0')
             self.set_metadata('version', mongo_version)
         except Exception:
             self.log.exception("Error when collecting the version from the mongo server.")
             mongo_version = '0.0'
 
-        tags = deepcopy(self.tags)
+        collector = ServerStatusCollector(self, self.db_name, tags, tcmalloc=self.collect_tcmalloc_metrics)
+        try:
+            collector.collect(cli)
+        except Exception:
+            self.service_check(SERVICE_CHECK_NAME, AgentCheck.CRITICAL, tags=self.service_check_tags)
+            raise
+        else:
+            self.service_check(SERVICE_CHECK_NAME, AgentCheck.OK, tags=self.service_check_tags)
+
+        collector = CurrentOpCollector(self, self.db_name, tags)
+        collector.collect(cli)
+
+        collector = DbStatCollector(self, self.db_name, tags)
+        collector.collect(cli)
+
         # Handle replica data, if any
         # See
         # http://www.mongodb.org/display/DOCS/Replica+Set+Commands#ReplicaSetCommands-replSetGetStatus  # noqa
-        if self.replica_check:
+        if self.replica_check and isinstance(self.deployment_type, ReplicaSetDeploymentType):
+            collector = ReplicaCollector(self, tags, self._previous_state)
```

Comment:
(Nit) IMO we should strive to pass any extra parameters (other than `check` and `tags`) using keyword arguments on all collector instantiations, to clarify that a/ they're optional and b/ what those arguments are from the collector's point of view (sometimes this differs from the variable name in this method's scope).

Eg here…

```suggestion
            collector = ReplicaCollector(self, tags, last_state=self._previous_state)
```